crates/bevy_ecs/src/bundle.rs

```diff
@@ -172,31 +195,32 @@ pub unsafe trait Bundle: DynamicBundle + Send + Sync + 'static {
         // Ensure that the `OwningPtr` is used correctly
         F: for<'a> FnMut(&'a mut T) -> OwningPtr<'a>,
         Self: Sized;
-
-    /// Registers components that are required by the components in this [`Bundle`].
-    fn register_required_components(
-        _components: &mut Components,
-        _storages: &mut Storages,
-        _required_components: &mut RequiredComponents,
-    );
 }
 
 /// The parts from [`Bundle`] that don't require statically knowing the components of the bundle.
 pub trait DynamicBundle {
+    /// An operation on the entity that happens _after_ inserting this bundle.
+    type Effect: BundleEffect;
     // SAFETY:
     // The `StorageType` argument passed into [`Bundle::get_components`] must be correct for the
     // component being fetched.
     //
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This passes
     /// ownership of the component values to `func`.
     #[doc(hidden)]
-    fn get_components(self, func: &mut impl FnMut(StorageType, OwningPtr<'_>));
+    fn get_components(self, func: &mut impl FnMut(StorageType, OwningPtr<'_>)) -> Self::Effect;
+}
+
+/// An operation on an [`Entity`] that occurs _after_ inserting a given [`Bundle`].
+/// See [`DynamicBundle::Effect`].
+pub trait BundleEffect {
```

Comment:
This seems fairly similar to an OnAdd ComponentHook.
I wonder if there could be an approach where we can define InternalComponentHooks that have access to EntityWorldMut instead of DeferredWorld.
Then the `SpawnRelatedBundle` would simply be a `SpawnRelatedComponent` component with an associated on_add hook to spawn children, etc.

We would get the same number of archetype moves as with the BundleEffect approach, no?
(apart maybe from the fact that this extra component would have to remove itself)

Reply:
Actually I do like the fact that this kind of unintuitive behavior (spawning extra entities) is only present on `Bundle`, which is an entity-creation-time abstraction instead of `Component` which is a post-creation-time abstraction

Reply:
Yeah I don't like the concept of inserting unnecessary components like (SpawnRelatedComponent) / prefer the bundle approach for this category of thing.

Reply:
> We would get the same number of archetype moves as with the BundleEffect approach, no?

It would require storing the list of related entities/bundles to spawn on the component (allocating space for it), then reading+removing it (internal to the component ... to avoid an archetype move ... but then it leaves that space allocated).